clients/venice-thin-client/src/main/java/com/linkedin/venice/client/schema/RouterBackedSchemaReader.java

```diff
@@ -300,6 +298,309 @@ private SchemaEntry fetchSingleSchema(String requestPath, boolean isValueSchema)
+   */
+  private SchemaEntry maybeFetchValueSchemaEntryById(int valueSchemaId, boolean forceRefresh) {
+    if (forceRefresh) {
+      SchemaEntry oldEntry = valueSchemaEntryMap.get(valueSchemaId);
+      // Do not need to refresh if the schema id is already present in the schema repo.
+      if (oldEntry != null && isValidSchemaEntry(oldEntry)) {
+        valueSchemaMapR.put(oldEntry.getSchema(), valueSchemaId);
+        return oldEntry;
+      }
+      SchemaEntry entry = fetchValueSchemaEntryFromRouter(valueSchemaId);
+      if (entry == null) {
+        valueSchemaEntryMap.put(valueSchemaId, NOT_EXIST_VALUE_SCHEMA_ENTRY);
+        return NOT_EXIST_VALUE_SCHEMA_ENTRY;
+      } else {
+        valueSchemaEntryMap.put(valueSchemaId, entry);
+        valueSchemaMapR.put(entry.getSchema(), valueSchemaId);
+        return entry;
+      }
+    } else {
+      return valueSchemaEntryMap.computeIfAbsent(valueSchemaId, (schemaId) -> {
+        SchemaEntry entry = fetchValueSchemaEntryFromRouter(valueSchemaId);
+        if (entry == null) {
+          return NOT_EXIST_VALUE_SCHEMA_ENTRY;
+        }
+        // Every time when we fetch a new value schema to cache during non-force-refresh logic, we should try to mark
+        // the flag as true.
+        shouldRefreshLatestValueSchemaEntry.compareAndSet(false, true);
+        valueSchemaMapR.put(entry.getSchema(), valueSchemaId);
+        return entry;
+      });
+    }
+  }
+
+  /**
+   * This method will fetch the latest update schema by ID, and potentially will update the cached schema map.
+   * If the schema:
+   * (1) Exists in the cached schema map, we will directly return it.
+   * (2) Does not exist in the cached schema map, we will perform one time single schema fetch to request it from router,
+   * and update the result in the cached schema map.
+   * (3) If the schema was queried previously and not found, we will:
+   * -- (a) Return not found, if forceRefresh flag == false.
+   * -- (b) Perform one time router query, update and return the result if forceRefresh flag == true.
+   */
+  private DerivedSchemaEntry maybeUpdateAndFetchUpdateSchemaEntryById(int valueSchemaId, boolean forceRefresh) {
+    if (forceRefresh) {
+      DerivedSchemaEntry derivedSchemaEntry = fetchUpdateSchemaEntryFromRouter(valueSchemaId);
+      if (derivedSchemaEntry == null) {
+        valueSchemaIdToUpdateSchemaEntryMap.put(valueSchemaId, NOT_EXIST_UPDATE_SCHEMA_ENTRY);
+        return NOT_EXIST_UPDATE_SCHEMA_ENTRY;
+      } else {
+        valueSchemaIdToUpdateSchemaEntryMap.put(valueSchemaId, derivedSchemaEntry);
+        return derivedSchemaEntry;
+      }
+    } else {
+      return valueSchemaIdToUpdateSchemaEntryMap.computeIfAbsent(valueSchemaId, (id) -> {
+        DerivedSchemaEntry derivedSchemaEntry = fetchUpdateSchemaEntryFromRouter(id);
+        if (derivedSchemaEntry == null) {
+          return NOT_EXIST_UPDATE_SCHEMA_ENTRY;
+        }
+        return derivedSchemaEntry;
```

Comment:
After the change, it seems `forceRefresh` param is useless since it will use the schema if exists and call Router endpoint if absent.
This comment applies to function: `maybeFetchValueSchemaEntryById` as well.